---
title: "RNA-seq analysis of cbf11, mga2, and cbf11 mga2 mutants in YES and YES+ammonium"
author: "Martin Převorovský, prevorov@natur.cuni.cz, http://www.GenoMik.cz"
output: html_document
---

This analysis utilizes data the following two experiments:
* 2020_RNA-seq_ammonium (WT, cbf11 in YES / YES+ammonium)
* 2022_RNA-seq_mga2 (mga2, mga2 cbf11 in YES / YES+ammonium)

```{r setup, message=FALSE}
library(rtracklayer)
library(GenomicAlignments)
library(DESeq2)
library(RColorBrewer)
library(pheatmap)

knitr::opts_chunk$set(echo = TRUE)

genome_dir <- './genome/'
genome_gff <- paste0(genome_dir, 'Schizosaccharomyces_pombe_all_chromosomes.gff3')
bam_dir1 <- './BAM_RNA-seq/'
bam_dir2 <- './../2020_RNA-seq_ammonium/BAM_RNA-seq/'
coverage_dir <- './coverage_RNA-seq/'
image_dir <- './images/'
deg_dir <- './DEG_RNA-seq/'
```

# Identify differentially expressed genes

## Generate gene-level counts tables.

* Raw count data are needed for DEG calling in `DESeq2`.

* TPM-normalized (Transcripts Per Million) count data are suitable for general comparisons of the samples.

```{r, eval=TRUE}
gff <- import.gff3(genome_gff)
gff.genes <- gff[gff$type == 'gene', ]

sample_info <- read.delim('sampleInfo_cbf11+mga2', stringsAsFactors = FALSE, header = TRUE, sep = '\t')
sample_info[, 'cultivation_run'] <- factor(sample_info[, 'cultivation_run'])
sample_info[, 'genotype'] <- factor(sample_info[, 'genotype'])
sample_info[, 'medium'] <- factor(sample_info[, 'medium'])
sample_info[, 'sequencing_run'] <- factor(sample_info[, 'sequencing_run'])

# mga2, mga2 cbf11
bam1 <- BamFileList(list.files(bam_dir1, 
                               pattern = '\\.bam$', 
                               full.names = TRUE),
                    asMates = FALSE,
                    yieldSize = 1000000)
names(bam1) <- sample_info$sample_ID[1:12]
se1 <- summarizeOverlaps(gff.genes,
                         bam1,
                         mode = 'Union',
                         singleEnd = TRUE,
                         ignore.strand = FALSE,
                         preprocess.reads = invertStrand)

# WT, cbf11
bam2 <- BamFileList(list.files(bam_dir2, 
                               pattern = '\\.bam$', 
                               full.names = TRUE),
                    asMates = TRUE,
                    yieldSize = 1000000)
names(bam2) <- sample_info$sample_ID[13:24]
se2 <- summarizeOverlaps(gff.genes,
                         bam2,
                         mode = 'Union',
                         singleEnd = FALSE,
                         fragments = FALSE,
                         ignore.strand = FALSE,
                         preprocess.reads = invertStrand)

se <- cbind(se1, se2)
rownames(se) <- gff.genes$ID
colData(se) <- cbind(colData(se), sample_info)
save(se, se1, se2, file = 'se_cbf11+mga2.rda')
#load(file = 'se_cbf11+mga2.rda')

counts <- assay(se)

# normalize to get TPM values (transcripts per million)
gene_widths <- width(gff.genes)
counts.norm <- counts
for (i in 1:nrow(counts.norm)){
  counts.norm[i, ] <- counts.norm[i, ] / gene_widths[i]
}
count_sums <- colSums(counts.norm) / 1000000
for (i in 1:length(count_sums)){
  counts.norm[, i] <- counts.norm[, i] / count_sums[i]
}

counts <- cbind(gff.genes$Name, counts)
counts.norm <- cbind(gff.genes$Name, counts.norm)
colnames(counts)[1] <- 'gene_name'
colnames(counts.norm)[1] <- 'gene_name'
write.csv(counts, file = paste0(coverage_dir, 'counts_cbf11+mga2.csv'), quote = FALSE)
write.csv(counts.norm, file = paste0(coverage_dir, 'counts_cbf11+mga2.TPM-normalized.csv'), quote = FALSE)
rm(gene_widths, count_sums)
```

## QC of count table data

```{r, eval=TRUE}
dds <- DESeqDataSet(se, design = ~ cultivation_run + medium + genotype)
dds <- DESeq(dds)
rld <- rlog(dds, blind = TRUE)

sampleDists <- dist(t(assay(rld)))
sampleDistMatrix <- as.matrix(sampleDists)
rownames(sampleDistMatrix) <- rld$sample_ID
colnames(sampleDistMatrix) <- NULL
colors <- colorRampPalette(rev(brewer.pal(9, "Blues")))(255)
pdf(paste0(image_dir, 'heatmap_RNA-seq_sampleDist_rlog_cbf11+mga2.pdf'))
pheatmap(sampleDistMatrix,
         clustering_distance_rows = sampleDists,
         clustering_distance_cols = sampleDists,
         col = colors)
dev.off()

pdf(paste0(image_dir, 'PCA_RNA-seq_rlog_cbf11+mga2.pdf'))
plotPCA(rld, intgroup = c('genotype', 'medium', 'cultivation_run'))
plotPCA(rld, intgroup = 'genotype')
plotPCA(rld, intgroup = 'medium')
plotPCA(rld, intgroup = 'cultivation_run')
dev.off()
```

* Heatmap results - samples cluster strongly by genotype. Possibly also by sequencing run (hard to tell as no genotype was present in both runs).

* PCA results - genotype is the strongest differentiating factor


## DEG calling

```{r, eval=TRUE}
# Benjamini-Hochberg p value adjustment (FDR)
padj_threshold <- 0.05

dir.create(deg_dir)

# mga2_cbf11-YES vs mga2-YES
dds <- DESeqDataSet(se, design = ~ cultivation_run + genotype)
dds <- dds[, dds$medium =='YES'] # remove YES+ammonium samples ('YES_AMM')
dds$medium <- droplevels(dds$medium) # remove factor levels without samples
dds$genotype <- relevel(dds$genotype, ref = 'mga2') # change order of factor levels to get mga2_cbf11/mga2 fold change (= define the reference)
dds <- DESeq(dds)
res <- results(dds, alpha = padj_threshold)
summary(res[res$padj <= padj_threshold & !is.na(res$padj), ])
res <- cbind(counts[, 'gene_name'], res)
colnames(res)[1] <- 'gene_name'
write.csv(res, file = paste0(deg_dir, 'DESeq2results_mga2_cbf11-YES_vs_mga2-YES.csv'), quote = FALSE)
res.sig <- res[which(res$padj <= padj_threshold), ]
res.sig <- res.sig[order(res.sig$log2FoldChange), ]
write.csv(res.sig, file = paste0(deg_dir, 'DESeq2results_mga2_cbf11-YES_vs_mga2-YES.SIG.csv'), quote = FALSE)

# mga2_cbf11-YES+AMM vs mga2-YES+AMM
dds <- DESeqDataSet(se, design = ~ cultivation_run + genotype)
dds <- dds[, dds$medium =='YES_AMM']
dds$medium <- droplevels(dds$medium)
dds$genotype <- relevel(dds$genotype, ref = 'mga2')
dds <- DESeq(dds)
res <- results(dds, alpha = padj_threshold)
summary(res[res$padj <= padj_threshold & !is.na(res$padj), ])
res <- cbind(counts[, 'gene_name'], res)
colnames(res)[1] <- 'gene_name'
write.csv(res, file = paste0(deg_dir, 'DESeq2results_mga2_cbf11-YES+AMM_vs_mga2-YES+AMM.csv'), quote = FALSE)
res.sig <- res[which(res$padj <= padj_threshold), ]
res.sig <- res.sig[order(res.sig$log2FoldChange), ]
write.csv(res.sig, file = paste0(deg_dir, 'DESeq2results_mga2_cbf11-YES+AMM_vs_mga2-YES+AMM.SIG.csv'), quote = FALSE)

# mga2-YES+AMM vs mga2-YES
dds <- DESeqDataSet(se, design = ~ cultivation_run + medium)
dds <- dds[, dds$genotype =='mga2']
dds$genotype <- droplevels(dds$genotype)
dds$medium <- relevel(dds$medium, ref = 'YES')
dds <- DESeq(dds)
res <- results(dds, alpha = padj_threshold)
summary(res[res$padj <= padj_threshold & !is.na(res$padj), ])
res <- cbind(counts[, 'gene_name'], res)
colnames(res)[1] <- 'gene_name'
write.csv(res, file = paste0(deg_dir, 'DESeq2results_mga2-YES+AMM_vs_mga2-YES.csv'), quote = FALSE)
res.sig <- res[which(res$padj <= padj_threshold), ]
res.sig <- res.sig[order(res.sig$log2FoldChange), ]
write.csv(res.sig, file = paste0(deg_dir, 'DESeq2results_mga2-YES+AMM_vs_mga2-YES.SIG.csv'), quote = FALSE)

# mga2_cbf11-YES+AMM vs mga2_cbf11-YES
dds <- DESeqDataSet(se, design = ~ cultivation_run + medium)
dds <- dds[, dds$genotype =='mga2_cbf11']
dds$genotype <- droplevels(dds$genotype)
dds$medium <- relevel(dds$medium, ref = 'YES')
dds <- DESeq(dds)
res <- results(dds, alpha = padj_threshold)
summary(res[res$padj <= padj_threshold & !is.na(res$padj), ])
res <- cbind(counts[, 'gene_name'], res)
colnames(res)[1] <- 'gene_name'
write.csv(res, file = paste0(deg_dir, 'DESeq2results_mga2_cbf11-YES+AMM_vs_mga2_cbf11-YES.csv'), quote = FALSE)
res.sig <- res[which(res$padj <= padj_threshold), ]
res.sig <- res.sig[order(res.sig$log2FoldChange), ]
write.csv(res.sig, file = paste0(deg_dir, 'DESeq2results_mga2_cbf11-YES+AMM_vs_mga2_cbf11-YES.SIG.csv'), quote = FALSE)





# cbf11-YES vs mga2-YES
dds <- DESeqDataSet(se, design = ~ cultivation_run + genotype)
dds <- dds[, dds$medium =='YES'] # remove YES+ammonium samples ('YES_AMM')
dds <- dds[, dds$genotype %in% c('cbf11', 'mga2')]
dds$medium <- droplevels(dds$medium) # remove factor levels without samples
dds$genotype <- droplevels(dds$genotype) # remove factor levels without samples
dds$genotype <- relevel(dds$genotype, ref = 'mga2') # change order of factor levels to get cbf11/mga2 fold change (= define the reference)
dds <- DESeq(dds)
res <- results(dds, alpha = padj_threshold)
summary(res[res$padj <= padj_threshold & !is.na(res$padj), ])
res <- cbind(counts[, 'gene_name'], res)
colnames(res)[1] <- 'gene_name'
write.csv(res, file = paste0(deg_dir, 'DESeq2results_cbf11-YES_vs_mga2-YES.csv'), quote = FALSE)
res.sig <- res[which(res$padj <= padj_threshold), ]
res.sig <- res.sig[order(res.sig$log2FoldChange), ]
write.csv(res.sig, file = paste0(deg_dir, 'DESeq2results_cbf11-YES_vs_mga2-YES.SIG.csv'), quote = FALSE)
```


***

```{r}
sessionInfo()
```
